pkg/storageops/vsphere/vsphere.go

```diff
@@ -0,0 +1,378 @@
+	if err != nil {
+		return nil, err
+	}
+
+	logrus.Infof("Using following configuration for vsphere:")
+	logrus.Infof("  vCenter: %s:%s", cfg.VCenterIP, cfg.VCenterPort)
+	logrus.Infof("  Datacenter: %s", vmObj.Datacenter.Name())
+	logrus.Infof("  VMUUID: %s", cfg.VMUUID)
+
+	return &vsphereOps{
+		cfg:  cfg,
+		vm:   vmObj,
+		conn: vSphereConn,
+	}, nil
+}
+
+// Name returns name of the storage operations driver
+func (ops *vsphereOps) Name() string {
+	return "vsphere"
+}
+
+func (ops *vsphereOps) Create(opts interface{}, labels map[string]string) (interface{}, error) {
+	volumeOptions, ok := opts.(*vclib.VolumeOptions)
+	if !ok {
+		return nil, fmt.Errorf("invalid volume options specified to create: %v", opts)
+	}
+
+	if len(volumeOptions.Tags) == 0 {
+		volumeOptions.Tags = labels
+	} else {
+		for k, v := range labels {
+			volumeOptions.Tags[k] = v
+		}
+	}
+
+	if len(volumeOptions.Datastore) == 0 {
+		return nil, fmt.Errorf("datastore is required for the create call")
+	}
+
+	datastore := strings.TrimSpace(volumeOptions.Datastore)
+	logrus.Infof("Using datastore: %s for new disk", datastore)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	vmObj, err := ops.renewVM(ctx, ops.vm)
+	if err != nil {
+		return nil, err
+	}
+
+	ds, err := vmObj.Datacenter.GetDatastoreByName(ctx, datastore)
+	if err != nil {
+		logrus.Errorf("Failed to get datastore: %s due to: %v", datastore, err)
+		return nil, err
+	}
+
+	volumeOptions.Datastore = datastore
+
+	diskBasePath := filepath.Clean(ds.Path(diskDirectory)) + "/"
+	err = ds.CreateDirectory(ctx, diskBasePath, false)
+	if err != nil && err != vclib.ErrFileAlreadyExist {
+		logrus.Errorf("Cannot create dir %#v. err %s", diskBasePath, err)
+		return nil, err
+	}
+
+	diskPath := diskBasePath + volumeOptions.Name + ".vmdk"
+	disk := diskmanagers.VirtualDisk{
+		DiskPath:      diskPath,
+		VolumeOptions: volumeOptions,
+	}
+
+	diskPath, err = disk.Create(ctx, ds)
+	if err != nil {
+		logrus.Errorf("Failed to create a vsphere volume with volumeOptions: %+v on "+
+			"datastore: %s. err: %+v", volumeOptions, datastore, err)
+		return nil, err
+	}
+
+	// Get the canonical path for the volume path.
+	canonicalVolumePath, err := getCanonicalVolumePath(ctx, vmObj.Datacenter, diskPath)
+	if err != nil {
+		logrus.Errorf("Failed to get canonical vsphere disk path for: %s with "+
+			"volumeOptions: %+v on datastore: %s. err: %+v", diskPath, volumeOptions, datastore, err)
+		return nil, err
+	}
+
+	return canonicalVolumePath, nil
+
+}
+
+func (ops *vsphereOps) GetDeviceID(diskPath interface{}) (string, error) {
+	id, ok := diskPath.(string)
+	if !ok {
+		return "", fmt.Errorf("invalid input: %v to GetDeviceID", diskPath)
+	}
+
+	return id, nil
+}
+
+// Attach takes in the path of the vmdk file and returns where it is attached inside the vm instance
+func (ops *vsphereOps) Attach(diskPath string) (string, error) {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Q: Why run WithCancel() we are not cancelling or have a chain of ctx's. Maybe use WithTimeout() instead?

Reply:
We discussed this offline a bit. I'm going to leave this like this for now. The official examples for WithCancel() also use background (https://golang.org/pkg/context/#WithCancel)